src/main/java/games/strategy/engine/framework/GameRunner.java

```diff
@@ -261,10 +262,23 @@ public static void showMainFrame() {
 
       ProAI.gameOverClearCache();
       new Thread(() -> {
-        gameSelectorModel.loadDefaultGame(false);
-        final String fileName = System.getProperty(GameRunner.TRIPLEA_GAME_PROPERTY, "");
-        if (fileName.length() > 0) {
-          gameSelectorModel.load(new File(fileName), mainFrame);
+        final AtomicReference<WaitDialog> dialogReference = new AtomicReference<>();
```

Comment:
Not sure if this was your intention, but by using a modal dialog, we are effectively blocking the UI.  If that _was_ the intent, then a lot of this thread ping-ponging can be simplified by using `BackgroundTaskRunner`:

```diff
diff --git a/src/main/java/games/strategy/engine/framework/GameRunner.java b/src/main/java/games/strategy/engine/framework/GameRunner.java
index 0c23285a6..9cc82d25f 100644
--- a/src/main/java/games/strategy/engine/framework/GameRunner.java
+++ b/src/main/java/games/strategy/engine/framework/GameRunner.java
@@ -17,7 +17,6 @@ import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
 import java.util.Optional;
-import java.util.concurrent.atomic.AtomicReference;
 
 import javax.annotation.Nullable;
 import javax.swing.JFileChooser;
@@ -46,6 +45,7 @@ import games.strategy.engine.framework.startup.ui.ServerSetupPanel;
 import games.strategy.engine.framework.system.HttpProxy;
 import games.strategy.engine.framework.systemcheck.LocalSystemChecker;
 import games.strategy.engine.framework.ui.SaveGameFileChooser;
+import games.strategy.engine.framework.ui.background.BackgroundTaskRunner;
 import games.strategy.engine.framework.ui.background.WaitDialog;
 import games.strategy.engine.lobby.server.GameDescription;
 import games.strategy.net.Messengers;
@@ -261,30 +261,8 @@ public class GameRunner {
       SwingComponents.addWindowClosingListener(mainFrame, GameRunner::exitGameIfFinished);
 
       ProAI.gameOverClearCache();
-      new Thread(() -> {
-        final AtomicReference<WaitDialog> dialogReference = new AtomicReference<>();
-        SwingUtilities.invokeLater(() -> {
-          dialogReference.set(new WaitDialog(mainFrame, "Loading game..."));
-          dialogReference.get().setVisible(true);
-        });
 
-        try {
-          gameSelectorModel.loadDefaultGame(false);
-          final String fileName = System.getProperty(GameRunner.TRIPLEA_GAME_PROPERTY, "");
-          if (fileName.length() > 0) {
-            gameSelectorModel.load(new File(fileName), mainFrame);
-          }
-        } finally {
-          SwingUtilities.invokeLater(() -> {
-            dialogReference.get().setVisible(false);
-            dialogReference.get().dispose();
-          });
-        }
-        final String downloadableMap = System.getProperty(GameRunner.TRIPLEA_MAP_DOWNLOAD_PROPERTY, "");
-        if (!downloadableMap.isEmpty()) {
-          SwingUtilities.invokeLater(() -> DownloadMapsWindow.showDownloadMapsWindowAndDownload(downloadableMap));
-        }
-      }).start();
+      loadGame();
 
       if (System.getProperty(GameRunner.TRIPLEA_SERVER_PROPERTY, "false").equals("true")) {
         setupPanelModel.showServer(mainFrame);
@@ -294,6 +272,27 @@ public class GameRunner {
     });
   }
 
+  private static void loadGame() {
+    try {
+      BackgroundTaskRunner.runInBackgroundAndReturn("Loading game...", () -> {
+        gameSelectorModel.loadDefaultGame(false);
+        final String fileName = System.getProperty(GameRunner.TRIPLEA_GAME_PROPERTY, "");
+        if (fileName.length() > 0) {
+          gameSelectorModel.load(new File(fileName), mainFrame);
+        }
+
+        final String downloadableMap = System.getProperty(GameRunner.TRIPLEA_MAP_DOWNLOAD_PROPERTY, "");
+        if (!downloadableMap.isEmpty()) {
+          SwingUtilities.invokeLater(() -> DownloadMapsWindow.showDownloadMapsWindowAndDownload(downloadableMap));
+        }
+
+        return null;
+      });
+    } catch (final InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
   private static void checkLocalSystem() {
     final LocalSystemChecker localSystemChecker = new LocalSystemChecker();
     final Collection<Exception> exceptions = localSystemChecker.getExceptions();
```

Also, if you're going to block anyway, it seems that you can move the download map check back to the original code path:

```diff
diff --git a/src/main/java/games/strategy/engine/framework/GameRunner.java b/src/main/java/games/strategy/engine/framework/GameRunner.java
index 0c23285a6..68837f716 100644
--- a/src/main/java/games/strategy/engine/framework/GameRunner.java
+++ b/src/main/java/games/strategy/engine/framework/GameRunner.java
@@ -17,7 +17,6 @@ import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
 import java.util.Optional;
-import java.util.concurrent.atomic.AtomicReference;
 
 import javax.annotation.Nullable;
 import javax.swing.JFileChooser;
@@ -46,6 +45,7 @@ import games.strategy.engine.framework.startup.ui.ServerSetupPanel;
 import games.strategy.engine.framework.system.HttpProxy;
 import games.strategy.engine.framework.systemcheck.LocalSystemChecker;
 import games.strategy.engine.framework.ui.SaveGameFileChooser;
+import games.strategy.engine.framework.ui.background.BackgroundTaskRunner;
 import games.strategy.engine.framework.ui.background.WaitDialog;
 import games.strategy.engine.lobby.server.GameDescription;
 import games.strategy.net.Messengers;
@@ -261,30 +261,13 @@ public class GameRunner {
       SwingComponents.addWindowClosingListener(mainFrame, GameRunner::exitGameIfFinished);
 
       ProAI.gameOverClearCache();
-      new Thread(() -> {
-        final AtomicReference<WaitDialog> dialogReference = new AtomicReference<>();
-        SwingUtilities.invokeLater(() -> {
-          dialogReference.set(new WaitDialog(mainFrame, "Loading game..."));
-          dialogReference.get().setVisible(true);
-        });
 
-        try {
-          gameSelectorModel.loadDefaultGame(false);
-          final String fileName = System.getProperty(GameRunner.TRIPLEA_GAME_PROPERTY, "");
-          if (fileName.length() > 0) {
-            gameSelectorModel.load(new File(fileName), mainFrame);
-          }
-        } finally {
-          SwingUtilities.invokeLater(() -> {
-            dialogReference.get().setVisible(false);
-            dialogReference.get().dispose();
-          });
-        }
-        final String downloadableMap = System.getProperty(GameRunner.TRIPLEA_MAP_DOWNLOAD_PROPERTY, "");
-        if (!downloadableMap.isEmpty()) {
-          SwingUtilities.invokeLater(() -> DownloadMapsWindow.showDownloadMapsWindowAndDownload(downloadableMap));
-        }
-      }).start();
+      loadGame();
+
+      final String downloadableMap = System.getProperty(GameRunner.TRIPLEA_MAP_DOWNLOAD_PROPERTY, "");
+      if (!downloadableMap.isEmpty()) {
+        DownloadMapsWindow.showDownloadMapsWindowAndDownload(downloadableMap);
+      }
 
       if (System.getProperty(GameRunner.TRIPLEA_SERVER_PROPERTY, "false").equals("true")) {
         setupPanelModel.showServer(mainFrame);
@@ -294,6 +277,22 @@ public class GameRunner {
     });
   }
 
+  private static void loadGame() {
+    try {
+      BackgroundTaskRunner.runInBackgroundAndReturn("Loading game...", () -> {
+        gameSelectorModel.loadDefaultGame(false);
+        final String fileName = System.getProperty(GameRunner.TRIPLEA_GAME_PROPERTY, "");
+        if (fileName.length() > 0) {
+          gameSelectorModel.load(new File(fileName), mainFrame);
+        }
+
+        return null;
+      });
+    } catch (final InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
   private static void checkLocalSystem() {
     final LocalSystemChecker localSystemChecker = new LocalSystemChecker();
     final Collection<Exception> exceptions = localSystemChecker.getExceptions();
```

Bonus points if you want to add a new method to `BackgroundTaskRunner` that accepts a `Runnable` so you can get rid of the ugly `return null;` at the end of the background task lambda.  That would look something like:

```diff
diff --git a/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java b/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java
index 878a31894..0da437a48 100644
--- a/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java
+++ b/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java
@@ -22,6 +22,30 @@ public final class BackgroundTaskRunner {
   private BackgroundTaskRunner() {}
 
   /**
+   * Runs the specified action in the background without blocking the UI.
+   *
+   * <p>
+   * {@code backgroundAction} is run on a background (non-UI) thread. While the background action is running, a dialog
+   * is displayed with the specified message. When the background action is complete, the dialog is removed.
+   * </p>
+   *
+   * @param message The message displayed to the user while the background action is running.
+   * @param backgroundAction The action to run in the background.
+   *
+   * @throws IllegalStateException If this method is not called from the EDT.
+   * @throws InterruptedException If the UI thread is interrupted while waiting for the background action to complete.
+   */
+  public static void runInBackgroundAndConsume(
+      final String message,
+      final Runnable backgroundAction)
+      throws InterruptedException {
+    runInBackgroundAndReturn(message, () -> {
+      backgroundAction.run();
+      return null;
+    });
+  }
+
+  /**
    * Runs the specified action in the background without blocking the UI and returns the result of that action.
    *
    * <p>
```

Reply:
@ssoloff Your suggestion works just fine, but I'm not quite satisfied with the code.
I'm not entirely sure what you mean by "blocking the UI": Do you mean the EDT?
If so, I don't quite see where it is being blocked. We _are_ calling Dialog#setVisible, but the EDT manages to keep the rest of the EDT running.

In any case, I want to have (and use) an asynchonous method in the BackgroundTaskRunner class. Waiting for a method on another thread to execute is really not neccessary in this case...

Reply:
What I mean by "blocking the UI" is that the user can take no action.  For example, in the current `master`, I can click **Download Maps**, **Play Online**, or **Engine Preferences** and do some other work while the game is loading.  That's no longer possible once you display a modal dialog.

Reply:
Ahh, that's what you meant.
This is not entirely true.
Users are still able to press x

Reply:
Yeah, I wonder if we should disable that behavior (or at least provide the ability to do so) because most operations that use `BackgroundTaskRunner` really shouldn't allow the user to continue using the UI until the background task completes.

Reply:
@ssoloff Probably out of scope for this PR...
But if you can come up with a good solution, go ahead and submit a PR :)